waiter/src/waiter/descriptor.clj

```diff
@@ -405,17 +412,44 @@
-      (loop [router-id->success? {}
+                                                            (async/<! (await-service-goal-fallback-state-locally
+                                                                        fallback-state-atom service-id timeout sleep-duration goal))))}))]
+      (loop [router-id->result {}
              [[router-id response-chan] & remaining] (seq router-id->response-chan)]
         (if (and router-id response-chan)
           (recur
-            (assoc router-id->success? router-id (some-> response-chan
-                                            async/<!
-                                            :body
-                                            async/<!
-                                            utils/try-parse-json
-                                            (get "success?")))
+            (assoc router-id->result router-id (some-> response-chan
+                                                       async/<!
+                                                       :body
+                                                       async/<!
+                                                       utils/try-parse-json
+                                                       walk/keywordize-keys))
             remaining)
-          router-id->success?)))))
+          router-id->result)))))
+
+(defn extract-service-state
+  "Extract the service state which is consistent on all routers. The ping-result determines what checks are made.
+  The value returned is a map {:healthy? :exists? :service-id :status}"
+  [router-id retrieve-service-status-label-fn fallback-state-atom make-inter-router-requests-async-fn service-id ping-result]
+  (go-try
+    (let [router-comm-timeout-ms 5000
+          sleep-duration-ms 100
+          reduce-results (fn [router-service-state-list]
+                           {:healthy? (every? :service-healthy? router-service-state-list)
+                            :exists? (every? :service-exists? router-service-state-list)})
+          get-final-fallback-state (fn [event]
+                                     (go-try
+                                       (->> (await-service-goal-fallback-state fallback-state-atom make-inter-router-requests-async-fn router-id service-id router-comm-timeout-ms sleep-duration-ms event)
+                                            <?
+                                            vals
+                                            reduce-results)))
+          final-fallback-state (case ping-result
+                                 :timed-out (do
+                                              (log/info "skipping inter router checks because ping timed out")
+                                              {:healthy? (service-healthy? @fallback-state-atom service-id)
+                                               :exists? (service-exists? @fallback-state-atom service-id)})
```

Comment:
We should use the same value for the fallback state to avoid any race issues.
Might be painful while debugging if the values do not agree.
```suggestion
                                              (let [fallback-state @fallback-state-atom]
                                                {:healthy? (service-healthy? fallback-state service-id)
                                                 :exists? (service-exists? fallback-state service-id)}))
```